### ISSUE TITLE
ArgumentNullException error

### ISSUE THREAD

**Issue by papafe** [realm/realm-dotnet#3094]
### Discussed in https://github.com/realm/realm-dotnet/discussions/3093

<div type='discussions-op-text'>

<sup>Originally posted by **Unreal852** November 13, 2022</sup>
Hello, i'm making an application with AvaloniaUI and i decided to use Realm as my database but when i bind my view to fields that belongs to an IRealmObject it throws ArgumentNullException. I created a separate project to replicate the issue https://github.com/Unreal852/RealmAvalonia
I did not create an issue because i don't know if i'm doing things wrong or not.</div>

**Comment by Unreal852** [comment]
I'll give more details about this. 

I have my [AvaloniaUI](https://github.com/AvaloniaUI/Avalonia) app using MVVM pattern, so i got 3 components, the **View**, the **ViewModel** and the **Model**

Here is the **View** :
```XAML
 <Grid RowDefinitions="Auto *">
        <Button Content="Add" Command="{Binding AddPlayer}"/>
        <ListBox Grid.Row="1" Items="{Binding Players}">
            <ListBox.ItemTemplate>
                <DataTemplate DataType="{x:Type models:Player}">
                    <StackPanel Orientation="Vertical">
                        <TextBlock Text="{Binding Name}" />
                        <Label Content="{Binding Experience}" />
                    </StackPanel>
                </DataTemplate>
            </ListBox.ItemTemplate>
        </ListBox> 
    </Grid>
```

Here is the **ViewModel** :

```C#
public class MainWindowViewModel : ReactiveObject
{
    public ObservableCollection<Player> Players { get; } = new();

    public void AddPlayer()
    {
        Players.Add(new Player { Name = "Unreal", Experience = 15.2 });
    }
}
```

And here is the **Model** :

```C#
public partial class Player : IRealmObject
{
    public string? Name       { get; set; }
    public double  Experience { get; set; }
}
```

I'm using Realm source generator so i implemented **IRealmObject** to my **Model**.
When i run this and then add a new **Player** instance to in my **ViewModel** it throws `System.ArgumentNullException: Value cannot be null. (Parameter 'source')`

![realm_avalonia](https://user-images.githubusercontent.com/5130243/201681723-651b0011-9815-4dbd-a023-3b4b2c3a2057.gif)

Now if i remove the **IRealmObject** interface from my model, everything runs fine.

![realm_avalonia_2](https://user-images.githubusercontent.com/5130243/201682170-5382fad1-e2c1-4f9e-a0d7-c885b354bf3e.gif)

Also, you can try by yourself with this [Github Repo](https://github.com/Unreal852/RealmAvalonia), just clone and run it.

**Comment by papafe** [comment]
Thanks for the additional information! We'll investigate and get back to you

**Comment by goldindan** [comment]
The same issue, if i go back to 10.17 and using RealObject instead of IReactObject is working good

**Comment by papafe** [comment]
@Unreal852 and @goldindan I think I found the reason for the issue, and I'm working on a solution. 
This is happening only with unmanaged objects, and so when a Realm object is used in bindings before it has been added to a realm. 
I can also see that in your code you are using `ObservableCollection`. When using Realm you actually don't need it, as you can bind directly to queries, and the UI will be changed automatically. 
For example, you can do something like this for your `MainViewModel`:

```csharp
public class MainWindowViewModel : ReactiveObject
{
    private Realm _realm;
    public IQueryable<Player> Players { get; }

    public MainWindowViewModel()
    {
        _realm = Realm.GetInstance();
        Players = _realm.All<Player>();
    }

    public void AddPlayer()
    {
        _realm.Write(() =>
        {
            _realm.Add(new Player { Name = "Unreal", Experience = 15.2 });
        });
    }
}
```

With this approach you don't need to worry about the `Players` collection as before, you just need to update the database and the UI will reflect the changes. You can also make more complicated queries, for instance:
```csharp
Players = _realm.All<Player>().Where( p => p.Age > 25);
```
This is the approach that we generally recommend developers, as it is the most idiomatic one, and developers don't need to worry about keeping collections updated. 
Another note. In order to make the code that I've shown work with Avalonia you need to use `ItemsControl` instead of `ListBox` in the view, as `ListBox` needs to be bound to a collection that implements `IList`, but `Players` does not. I am not an expert of Avalonia, so I am not sure of what would be the best approach, and why they need the collection to implement `IList` necessarily, but I hope this will be of help.

**Comment by Unreal852** [comment]
Hi, thanks for the quick fix. I didn't know we could bind directly on Realm collections this is nice.

**Comment by papafe** [comment]
@Unreal852 just as an additional note, you can take look at our [example project](https://github.com/realm/realm-dotnet/tree/main/examples/QuickJournal) if you want to have some other example of how to use bindings with Realm. The project uses Xamarin, but I think you can get the idea

**Comment by goldindan** [comment]
How can we fix the problem of binding an unmanaged IRealmObject to the view?

**Comment by papafe** [comment]
@goldindan this will be fixed in the next version (the fix is in review in PR #3097)
